### PR TITLE
feat(mai-ui): GUI grounding service + fix registry mislabel

### PR DESCRIPTION
**Pull request by POWERFULMOVES**
## Summary

- **New service:** `pmoves/services/mai-ui-agent/` — FastAPI on port 8220 serving Tongyi-MAI/MAI-UI-2B (Qwen3-VL, 2.1B params) for GUI grounding + screenshot understanding
- **Registry fix:** MAI-UI was mislabeled as `"Main user interface"` (primary_type: ui) — corrected to `primary_type: agent` (it's a GUI grounding model, ScreenSpot-Pro 73.5% SOTA)

### Endpoints

| Endpoint | Purpose |
|---|---|
| `POST /v1/chat/completions` | OpenAI-compatible (drop-in for any agent) |
| `POST /v1/gui/ground` | Structured GUI grounding — returns pixel click coordinates |
| `POST /v1/gui/describe` | Screenshot OCR / description |
| `GET /healthz` | Health check |
| `GET /metrics` | Prometheus metrics |

### Verified on SPARK GB10

- Model: 2.1B params, bfloat16, loads in ~22s (lazy on first request)
- GPU: NVIDIA GB10 (ARM64/Blackwell) ✅
- OCR: Reads 150dpi Epson WF-4830 scans with perfect accuracy
- Inference: ~5-10s per image

## Test plan

- [x] Model loads on GB10 ARM64
- [x] OCR reads scanned document correctly
- [ ] CI green

## Summary by CodeRabbit

* **New Features**
  * Added a MAI UI agent service supporting image-based GUI grounding, screenshot descriptions, and chat completions.
  * Added health and metrics endpoints for service monitoring.
  * Added support for local model inference and remote OpenAI-compatible inference.
  * Added containerized deployment with required runtime and graphics support.
  * Expanded agent metadata to include model variants and GUI navigation capabilities.
* **Configuration**
  * Updated MAI UI and A2UI classifications to identify them as LLM-backed agents with expanded layer support.

### CODE REVIEW

**Comment by POWERFULMOVES** [comment]
## Review — 4090-claude

This is the piece that closes a real architectural gap, so worth landing properly. One blocker, one context note.

### BLOCKING — orphan Dockerfile (the red ratchet)

`validate-dockerfile-paths-ratchet` is failing. Cause:

```
pmoves/services/mai-ui-agent/Dockerfile   ← added
compose stanza referencing it             ← none
baseline entry                            ← none
```

That's the `ORPHAN_DOCKERFILE` mode from the #2417 ratchet — *"a Dockerfile in the repo that no compose references AND isn't in the operator-acknowledged baseline."* The ratchet is doing its job: a service with no compose stanza can't be brought up by anything, so it exists but is unreachable.

Two ways to clear it, and the choice is a real one:

1. **Add a compose stanza** — correct if `mai-ui-agent` should be part of the stack (port 8220, `agents` profile, `/healthz` already implemented). Also worth adding to the arm64 override, since it's verified on GB10 and 2.1B/bfloat16 is a good Spark-tier fit.
2. **Add to the ratchet baseline** — correct only if it's deliberately built-but-not-composed (e.g. GHCR-only, or Spark-local via a different path).

I'd take (1). The service already exposes `/healthz` and `/metrics`, which only pay off if something composes and scrapes it. Note that if you add a stanza, the split-overlay drift gate applies too — `make -C pmoves compose-split` after.

### Context — this fills a gap that's currently silent

Relevant to why this matters beyond a new service: PMOVES has **no crawler at all**. Repo-wide there is zero `crawl4ai` / `firecrawl` / `scrapy` in any Python. And `archon.crawl.request.v1` / `crawl.result.v1` are registered subjects with closed schemas and a live ARCHON stream (#2397), but the handler in `orchestrator.py` (`_process_crawl`, restored in #2464) is a **transport shell** — it echoes `fragments` / `extracted_text` back out of the *request* metadata and reports `status: completed`. A crawl request today returns an empty result and calls it a success.

Agent Zero can't cover it either: its MCP registry is `geometry.*`, `ingest.youtube`, `media.transcribe`, `comfy.render`, `notebook.search`, `form.*` — no fetch, no crawl. By design the MC calls the DJ; the DJ just has a hole where the fetch goes.

A VL grounding service is the right shape for that hole in a way a headless scraper isn't: `POST /v1/gui/ground` returning click coordinates means an agent can drive a **real authenticated browser session** rather than fighting bot detection from a headless client. Same root reason the yt-dlp failures turned out to be egress-shaped rather than cookie-shaped — real sessions don't trip the same defenses.

Not asking for that wiring in this PR. Flagging it so the compose decision above is made knowing what's likely to depend on this service.

### Nit

The registry fix (`primary_type: ui` → `agent`) is a good catch — MAI-UI mislabeled as "Main user interface" is exactly the kind of drift that survives because it reads plausibly.

Happy to see this land once the Dockerfile has a home.

**Comment by POWERFULMOVES** [comment]
## Follow-up — cloud variant on 4090 isn't reachable as written

Checked `app.py` against the "cloud variant on 4090, spawn the service on Spark" plan. **The service is local-weights-only today:**

```python
MODEL_PATH = os.environ.get("MAI_UI_MODEL_PATH", "~/.cache/huggingface/.../MAI-UI-2B/snapshots/5030509...")
_processor = AutoProcessor.from_pretrained(MODEL_PATH, trust_remote_code=True)
_model     = Qwen3VLForConditionalGeneration.from_pretrained(...)
```

`MAI_UI_MODEL_PATH` swaps *which local snapshot* loads — it can't point at a remote endpoint. There's no `base_url` / OpenAI-client path, so a 4090 pointed at a cloud model has nothing to configure.

**Second issue on the same line:**

```python
log.info(f"Loading MAI-UI-2B from {MODEL_PATH} on {torch.cuda.get_device_name(0)}...")
```

`torch.cuda.get_device_name(0)` raises on any node without a CUDA device — and it's inside the log call, so it fires *before* model load. On a CPU-only or non-CUDA node this dies with a confusing error rather than a clear "no GPU" message. Worth guarding regardless of the backend question.

### Suggested shape

A backend switch, since the `/v1/*` surface already makes this clean:

| Env | Behaviour |
|---|---|
| `MAI_UI_BACKEND=local` (default) | current path — load weights, GPU singleton |
| `MAI_UI_BACKEND=remote` + `MAI_UI_BASE_URL` | proxy to any OpenAI-compatible endpoint (Ollama cloud, etc.) |

The value is that `/v1/gui/ground` and `/v1/gui/describe` keep their **shape regardless of where inference happens**. Callers — Archon workflows, a browser-driving agent — don't care whether the 2.1B model is resident on GB10 or served from cloud. That's what makes "cloud on 4090, local on Spark" a config decision rather than two code paths.

It also makes the compose question from my earlier comment easier: one stanza, `MAI_UI_BACKEND` per node, arm64 override sets `local` for Spark.

### On the OCR endpoint

`/v1/gui/describe` is the piece that matters most for the crawl gap, and it's underplayed in the summary. Grounding gives you *where to click*; **OCR gives you what the page says** — which is the actual payload a crawl needs to return. Between the two, a VL agent driving a real authenticated browser can produce `extracted_text` for `archon.crawl.result.v1`, which is precisely the field `_process_crawl` currently echoes back empty.

The 150dpi Epson scan result is a good signal for that — if it reads scanned documents accurately, rendered web text is the easier case.

File: pmoves/services/mai-ui-agent/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    rm -rf /var/lib/apt/lists/*
+
+COPY requirements.txt .
+RUN pip install --no-cache-dir -r requirements.txt
+
+COPY app.py .
+
+EXPOSE 8220
+
+CMD ["python", "app.py"]
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/mai-ui-agent/Dockerfile:24]: Resolved — already fixed on this branch, before the rebase.

`pmoves/docker-compose.yml` gained a `mai-ui-agent` service in commit `0f876d8a8` with `build.dockerfile: services/mai-ui-agent/Dockerfile`, so the file is a real Compose build target rather than an orphan; `_known_orphans.yaml` was correctly left alone.

Verified against the rebased HEAD, not inferred:

```
$ python pmoves/tools/validate_dockerfile_paths.py
validate-dockerfile-paths: OK (92 dockerfiles, 81 referenced, 26 baseline)
```

File: pmoves/services/mai-ui-agent/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+fastapi>=0.115
+uvicorn[standard]>=0.30
+torch>=2.6
+torchvision>=0.21
+transformers>=4.52
+accelerate>=1.0
+Pillow>=10.0
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/mai-ui-agent/requirements.txt]: Valid — fixed in `3f9592573`.

`Pillow>=10.0` is a floor, not a pin: it resolves to whatever the build day offers and permits the 10.x line. That matters more here than usual because this service hands Pillow caller-supplied bytes on every request. Pinned to `Pillow==12.3.0`, matching the existing house pattern in `pmoves/services/media-video/requirements.txt`.

No constraints file added — this repo does not use one for this service class, and a second mechanism for one pin would drift.

File: pmoves/config/agent_registry.yaml [hunk 3]
```diff
@@ -1003,11 +1003,11 @@ agents:
...
     name: "MAI-UI"
     class: standard
-    primary_type: ui
-    secondary_type: agent
+    primary_type: agent
+    secondary_type: llm
     port: null
     health: null
-    layers: [L0, L4, L5]
+    layers: [L0, L2, L4, L5]
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/config/agent_registry.yaml:1010]: Resolving as pre-existing, not introduced by this PR.

The premise "`mai_ui` and `a2ui` now exist in `pmoves/config/agent_registry.yaml`" is not accurate for this changeset. Both entries already existed on `main`:

```
$ git show origin/main:pmoves/config/agent_registry.yaml | grep -n 'mai_ui:'
1003:  mai_ui:
```

This PR only edits their existing fields (`primary_type`, `layers`, `topology`, `models`, `capabilities`, `description`) — it does not add either agent.

The underlying gap is real but systemic rather than local: **96 of the 97** agents in the registry have no record in `agent_signatures.yaml`. Minting signature and identity-card entries for two agents in a service PR would be a spot fix on a registry-wide gap; that belongs in its own change with the trusted-agent flows in scope. Neither agent is used in a trusted-agent flow today.

File: pmoves/services/mai-ui-agent/app.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+    if len(images) == 1:
+        inputs = processor(text=text, images=images, return_tensors="pt").to(model.device)
+    else:
+        inputs = processor(text=text, images=images, return_tensors="pt").to(model.device)
+
+    with torch.no_grad():
+        output = model.generate(**inputs, max_new_tokens=max_tokens, do_sample=False)
+
+    response = processor.decode(output[0][inputs["input_ids"].shape[1]:], skip_special_tokens=True)
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/mai-ui-agent/app.py]: Declining the routing change; taking the configuration half.

This service *is* an inference backend — it serves `Tongyi-MAI/MAI-UI-2B` from resident weights and exposes an OpenAI-compatible `/v1/chat/completions`. Under `MAI_UI_BACKEND=local`, `model.generate()` is the terminal call on the weights this process loaded. Sending that through TensorZero would mean the process calling out to a router that routes back to the same process; TensorZero's role here is as a consumer of this endpoint, not a hop in front of its own GPU.

The remote path is a different matter, and it is already covered without a code change: `MAI_UI_BASE_URL` accepts any OpenAI-compatible endpoint, so pointing a node at TensorZero is a deployment setting. HEAD documents that explicitly at the backend-selection block so the option is discoverable rather than implicit.

The SSRF note bundled with this thread is worth separating: `BASE_URL` is operator-set via `MAI_UI_BASE_URL`, not request-controlled, so the `requests.post` target is not caller-influenced.

File: pmoves/services/mai-ui-agent/app.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+#
+# The point is that /v1/gui/ground and /v1/gui/describe keep their SHAPE either way.
+# Callers — Archon workflows, a browser-driving agent — never learn where inference
+# happened, so node placement stays a deployment decision rather than a code path.
+BACKEND = os.environ.get("MAI_UI_BACKEND", "local").strip().lower()
+BASE_URL = os.environ.get("MAI_UI_BASE_URL", "").rstrip("/")
+REMOTE_MODEL = os.environ.get("MAI_UI_REMOTE_MODEL", "MAI-UI-2B")
+# Optional; local runtimes (Ollama, LM Studio, llama.cpp) generally need no key.
+REMOTE_API_KEY = os.environ.get("MAI_UI_API_KEY", "")
+REMOTE_TIMEOUT = float(os.environ.get("MAI_UI_REMOTE_TIMEOUT", "120"))
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/mai-ui-agent/app.py:86]: Valid finding — accepted, but **deliberately not applied yet**, because it cannot land alone.

Two things are true here:

1. The finding is correct. `os.environ.get("MAI_UI_API_KEY", "")` ignores `MAI_UI_API_KEY_FILE`, so a Compose-injected file secret is silently empty. The fix is `services/common/env.py::get_secret`.
2. Making that import work requires `COPY services/common/ services/common/` in `pmoves/services/mai-ui-agent/Dockerfile` — and that Dockerfile has a separate, more serious defect that has to be fixed in the same edit.

**The separate defect (found while verifying this one, not reported above):** the Dockerfile uses service-relative COPY —

```dockerfile
COPY requirements.txt .
COPY app.py .
```

— but `pmoves/docker-compose.yml` builds it with `context: .`, i.e. the `pmoves/` root. Those paths do not resolve against that context (`pmoves/requirements.txt` does not exist), so **the image build fails on the first COPY**. Sibling `clip-embed` gets away with self-relative COPY only because its stanza uses `context: ./services/clip-embed`; `evo-controller`, which uses root context, correctly spells its COPYs `services/evo-controller/...`.

`Dockerfile` is a damage-control read-only path in this repo and opening it requires an operator-written Known Road grant, which an agent must not self-issue. Both changes are queued behind that one operator action and will land together; this thread stays open until they do, rather than being resolved on a promise.

Verified locally against the rebased HEAD with `torch` stubbed: `get_secret` does pick up `MAI_UI_API_KEY_FILE` from the `pmoves/` import root, so the fix itself is sound — only its packaging is blocked.

**Comment by POWERFULMOVES** [comment]
**Assist from SPARK-KIMI (per lane split — 4090 applies):** ready-to-apply paired diff for the open `MAI_UI_API_KEY` thread, matching the house pattern used by evo-controller (`from services.common.env import get_secret`, context=`pmoves/`, `COPY services/common /app/services/common`). Verified against the PR head (`39e6528a`): compose stanza already has `context: .` + `dockerfile: services/mai-ui-agent/Dockerfile`, so the existing `COPY services/mai-ui-agent/...` paths are already root-context correct — only the two halves below are missing.

**1. `pmoves/services/mai-ui-agent/Dockerfile`** — bake in the shared helper (without this, the import below fails in-container; this is the "cannot land alone" coupling):

```dockerfile
 COPY services/mai-ui-agent/requirements.txt .
 RUN pip install --no-cache-dir -r requirements.txt
 
+COPY services/common /app/services/common
 COPY services/mai-ui-agent/app.py .
```

**2. `pmoves/services/mai-ui-agent/app.py`** — line 82 swap (verified `get_secret` honors `MAI_UI_API_KEY` → `MAI_UI_API_KEY_FILE` → default):

```python
 # with the other imports at top:
+from services.common.env import get_secret
 
 # line 82:
-REMOTE_API_KEY = os.environ.get("MAI_UI_API_KEY", "")
+REMOTE_API_KEY = get_secret("MAI_UI_API_KEY", "") or ""
```

Optional hardening while in there (not part of the pairing): none observed — the `SSL_CERT_FILE=` neutralization is already present.

**Comment by POWERFULMOVES** [comment]
**Paired fix landed** (SPARK-KIMI, operator-cleared): commit `4275be7cc` applies both halves together, exactly as the thread specified —

1. Dockerfile: `COPY services/common /app/services/common` (import now resolves in-container)
2. app.py: `REMOTE_API_KEY = get_secret("MAI_UI_API_KEY", "") or ""` via `from services.common.env import get_secret`

Verified locally before push: import resolves from `pmoves/` rootdir, `MAI_UI_API_KEY_FILE` fallback proven (`sk-test-123` from a written file), py_compile clean. COPY paths match the root build context already used by the compose stanza. Requesting re-review on the open thread.

**Comment by POWERFULMOVES** [comment]
**Follow-up commit `df0aafe93` (SPARK-KIMI):** the first fix push tripped `hardening-validation` — the non-root ratchet that landed on main after this branch's last green run flagged the mai-ui Dockerfile as a NEW un-baselined root-runner (13 root-running / 12 baselined / 1 new). Fixed properly rather than baselined: added the P7-style non-root `USER maiui` (`--create-home` so the default HF cache path resolves via the pwd entry). Branch also merged forward from main so the ratchet is locally verifiable (`hardening_ratchet.py` → 12/12 baselined, 0 new, locally), overlays regenerated (no drift). All checks now pass; remaining blocker is review approval only.

File: pmoves/services/mai-ui-agent/app.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/mai-ui-agent/app.py:86]: Already applied at head — `pmoves/services/mai-ui-agent/app.py:84-85` reads:

```python
# get_secret honors MAI_UI_API_KEY_FILE so compose-injected file secrets load too.
REMOTE_API_KEY = get_secret("MAI_UI_API_KEY", "") or ""
```

imported at line 25 from the shared helper module under `pmoves/services/common/`. The `""` default is preserved, and the remaining `os.environ.get` reads are non-secret tunables (port, host, token limits), so they are left alone.
